Cython/Utility/FunctionArguments.c

```diff
@@ -1,30 +1,63 @@
 //////////////////// ArgTypeTest.proto ////////////////////
 
 
+// Exact is 0 (False), 1 (True) or 2 (True and from annotation)
+// The latter gives a small amount of extra error diagnostics
 #define __Pyx_ArgTypeTest(obj, type, none_allowed, name, exact) \
     ((likely(__Pyx_IS_TYPE(obj, type) | (none_allowed && (obj == Py_None)))) ? 1 : \
         __Pyx__ArgTypeTest(obj, type, name, exact))
 
 static int __Pyx__ArgTypeTest(PyObject *obj, PyTypeObject *type, const char *name, int exact); /*proto*/
 
 //////////////////// ArgTypeTest ////////////////////
+//@substitute: naming
 
 static int __Pyx__ArgTypeTest(PyObject *obj, PyTypeObject *type, const char *name, int exact)
 {
     __Pyx_TypeName type_name;
     __Pyx_TypeName obj_type_name;
+    PyObject *extra_info = $empty_unicode;
+    int from_annotation_subclass = 0;
     if (unlikely(!type)) {
         PyErr_SetString(PyExc_SystemError, "Missing type object");
         return 0;
     }
     else if (!exact) {
         if (likely(__Pyx_TypeCheck(obj, type))) return 1;
+    } else if (exact == 2) {
+        // type from annotation
+        if (__Pyx_TypeCheck(obj, type)) {
+            from_annotation_subclass = 1;
+            extra_info = PYUNICODE("Note that Cython is deliberately stricter than PEP-484 and rejects subclasses of builtin types. If you need to pass subclasses then set the 'annotation_typing' directive to False.");
```

Comment:
We could also suggest using fused types as an alternative, but most cases would probably best be served with a plain `object` type, i.e. no specific type annotation.

Reply:
It's a bit of a guess, but I'm assuming the people getting caught by this are using annotations for "human readability" reasons. So they're probably happier with Cython ignoring the annotations than having to do something complicated with fused types.